node/core/pvf/src/host.rs

```diff
@@ -0,0 +1,1049 @@
+	let (to_host_tx, to_host_rx) = mpsc::channel(10);
+
+	let validation_host = ValidationHost {
+		to_host_tx: Mutex::new(to_host_tx),
+	};
+
+	let (to_prepare_pool, from_prepare_pool, run_prepare_pool) = prepare::start_pool(
+		config.prepare_worker_program_path.to_owned(),
+		config.prepare_worker_spawn_timeout,
+	);
+
+	let (to_prepare_queue_tx, from_prepare_queue_rx, run_prepare_queue) = prepare::start_queue(
+		config.prepare_workers_soft_max_num,
+		config.prepare_workers_hard_max_num,
+		config.cache_path.clone(),
+		to_prepare_pool,
+		from_prepare_pool,
+	);
+
+	let (to_execute_queue_tx, run_execute_queue) = execute::start(
+		config.execute_worker_program_path.to_owned(),
+		config.execute_workers_max_num,
+		config.execute_worker_spawn_timeout,
+	);
+
+	let (to_sweeper_tx, to_sweeper_rx) = mpsc::channel(100);
+	let run_sweeper = sweeper_task(to_sweeper_rx);
+
+	let run = async move {
+		let artifacts = Artifacts::new(&config.cache_path).await;
+
+		futures::pin_mut!(
+			run_prepare_queue,
+			run_prepare_pool,
+			run_execute_queue,
+			run_sweeper
+		);
+
+		run(
+			Inner {
+				cache_path: config.cache_path,
+				cleanup_pulse_interval: Duration::from_secs(3600),
+				artifact_ttl: Duration::from_secs(3600 * 24),
+				artifacts,
+				to_host_rx,
+				to_prepare_queue_tx,
+				from_prepare_queue_rx,
+				to_execute_queue_tx,
+				to_sweeper_tx,
+				awaiting_prepare: AwaitingPrepare::default(),
+			},
+			run_prepare_pool,
+			run_prepare_queue,
+			run_execute_queue,
+			run_sweeper,
+		)
+		.await
+	};
+
+	(validation_host, run)
+}
+
+/// An execution request that should execute the PVF (known in the context) and send the results
+/// to the given result sender.
+#[derive(Debug)]
+struct PendingExecutionRequest {
+	params: Vec<u8>,
+	result_tx: ResultSender,
+}
+
+/// A mapping from an artifact ID which is in preparation state to the list of pending exeuction
+/// requests that should be executed once the artifact's prepration is finished.
+#[derive(Default)]
+struct AwaitingPrepare(HashMap<ArtifactId, Vec<PendingExecutionRequest>>);
+
+impl AwaitingPrepare {
+	fn add(
+		&mut self,
+		artifact_id: ArtifactId,
+		params: Vec<u8>,
+		result_tx: ResultSender,
+	) {
+		self.0
+			.entry(artifact_id)
+			.or_default()
+			.push(PendingExecutionRequest { params, result_tx });
+	}
+
+	fn take(&mut self, artifact_id: &ArtifactId) -> Vec<PendingExecutionRequest> {
+		self.0.remove(artifact_id).unwrap_or_default()
+	}
+}
+
+struct Inner {
+	cache_path: PathBuf,
+	cleanup_pulse_interval: Duration,
+	artifact_ttl: Duration,
+	artifacts: Artifacts,
+
+	to_host_rx: mpsc::Receiver<ToHost>,
+
+	to_prepare_queue_tx: mpsc::Sender<prepare::ToQueue>,
+	from_prepare_queue_rx: mpsc::UnboundedReceiver<prepare::FromQueue>,
+
+	to_execute_queue_tx: mpsc::Sender<execute::ToQueue>,
+	to_sweeper_tx: mpsc::Sender<PathBuf>,
+
+	awaiting_prepare: AwaitingPrepare,
+}
+
+#[derive(Debug)]
+struct Fatal;
+
+async fn run(
+	Inner {
```

Comment:
So this only exists to have named parameters? But why do we pass 4 further parameters below?

Reply:
again, historical reasons )

I had to extract these task-futures just because I didn't want to name them, so passing them as parameters allowed me to use `impl Future`.